Improvements of AbstractActor API, #21717

patriknw opened the pull request:
Only a prototype to spark discussion so far. Should we do some final API changes to AbstractActor when making it non-experimental in 2.5?

There are a few things with the `AbstractActor` and the `ReceiveBuilder` that have been annoying me.

* would be better with an abstract method instead of setting the `receive` from constructor
  * gives compilation error instead of runtime exception if forgotten
  * defining receive deserves a better location than the constructor (subjective, I guess)
* avoid scary types, `PartialFunction`, `BoxedUnit`, `UnitPFBuilder<I>`

The changes are not source compatible but the migration steps are rather straight forward. Not sure if it would upset users?

It would be great if we could deprecate `UntypedActor`. I added `UntypedAbstractActor` as a migration path for those who want to continue with `onMessage(Object msg)`.

I have not converted all example yet, but enough to get a feeling for it.

patriknw committed:
improve AbstractActor, #21717

* Receive class that wraps PartialFunction, to avoid
  scary scala types
* move AbstractActorContext to AbstractActor.ActorContext
* converting docs, many, many UntypedActor
* removing UntypedActor docs
* add unit test for ReceiveBuilder
* MiMa filters
* consistent use of getContext(), self(), sender()
* rename cross references
* migration guide
* skip samples for now
* improve match type safetyi, add matchUnchecked
  * the `? extends P` caused code like this to compile:
    `match(String.class, (Integer i) -> {})`
  * added matchUnchecked, since it can still be useful (um, convenient)
    to be able to do:
    `matchUnchecked(List.class, (List<String> list) -> {})`
* eleminate some scala.Option
  * preRestart
  * findChild
  * ActorIdentity.getActorRef

patriknw commented:
@jypma what is your thoughts on this PR? (read description at the top)

guidomedina commented:
I have an `AbstractSupervisor` which extends `AbstractActor`, most of the times my `receive(...)` from my `AbstractSupervisor` is enough but few times I need to replace it, so super-classes has to `become(...)` instead of `receive(...)` at the constructor level.

I'm not sure what to suggest to be honest, I'm just giving one use case which is kind of weird.

patriknw commented:
Good input, @guidomedina 
Do I understand it correctly that then you could override the `initalReceive` method when needed.

It's also possible to use `become` from `preStart`.

guidomedina commented:
Object Oriented wise you won't have to `become(...)` when instantiating, that's my problem right now, by overwritting you don't have to initially `become(...)`

Semantically speaking when I instantiate my actor, it has just being created, if its superclass has invoked `receive(...)` in the constructor then the subclass will have to `become(...)`, this way makes more sense because `receive(...)` is overwitten hence using `become(..)` only when you are really changing, a true state machine and not a hack.

guidomedina commented:
Edited previous answer.

jypma commented:
In my opinion, there's nothing wrong with the "scary type" `PartialFunction`. It's easily explained what it does. 

We use `PartialFunction.orElse` all the time in our Java code, to separate out shared behaviour. I've also used the stacking behaviour the Java abstract actor subclasses I've introduced, e.g. for replication. I'd be sad to lose that ability from Java. Example [here](https://github.com/Tradeshift/ts-reaktive/blob/master/ts-reaktive-actors/src/main/java/com/tradeshift/reaktive/actors/AbstractStatefulPersistentActor.java#L73) which delegates specific types to [here](https://github.com/Tradeshift/ts-reaktive/blob/master/ts-reaktive-actors/src/main/java/com/tradeshift/reaktive/actors/AbstractStatefulPersistentActor.java#L103).

If the intent is to simplify the type signatures, we should introduce a proper (immutable and hence reusable, chainable) java `Receive` type, not just bump the `ReceiveBuilder` to top level because it happens to already exist.

Since I'm currently ill, I will be able to provide only limited feedback on this.

ktoso commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -46,29 +50,97 @@ object AbstractActor {
+   * up of resources before Actor is terminated.
+   */
+  @throws(classOf[Exception])
+  override def preRestart(reason: Throwable, message: Option[Any]): Unit = super.preRestart(reason, message)
```
why do we need these overrides here? Don't seem to change any types?

johanandren replied:
Should it be `Optional[Any]` instead here?

patriknw replied:
would be nice, might work to delegate to such a signature, the delegations between those hooks are crazy though. I'll give it a thought.

johanandren replied:
No change here?

johanandren replied:
This was the one I was thinking about, didn't remember I actually had commented on it

patriknw replied:
This should be possible to change, by delegating to a new, and deprecating this one. Eventually we can make it `final`.

ktoso commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -46,29 +50,97 @@ object AbstractActor {
+   * case of an unhandled [[akka.actor.Terminated]] message) or publishes an [[akka.actor.UnhandledMessage]]
+   * to the actor's system's [[akka.event.EventStream]].
+   */
+  override def unhandled(message: Any): Unit = super.unhandled(message)
```
ok here we override to change docs I assume? Good.

ktoso commented on akka-cluster-sharding/src/test/java/akka/cluster/sharding/ClusterShardingTest.java:
```diff
@@ -187,30 +188,30 @@ else if (msg.equals(ReceiveTimeout.getInstance()))
-        }).
-        match(String.class, s -> s.equals("stop-system"), s -> {
+        })
+        .match(String.class, s -> s.equals("stop-system"), s -> {
```
👍

ktoso commented on akka-actor/src/main/java/akka/japi/pf/ReceiveBuilder.java:
```diff
@@ -32,78 +35,147 @@
-  public static <P> UnitPFBuilder<Object> match(final Class<? extends P> type, FI.UnitApply<? extends P> apply) {
-    return UnitMatch.match(type, apply);
+  @SuppressWarnings("unchecked")
+  public <P> ReceiveBuilder match(final Class<? extends P> type, final FI.UnitApply<? extends P> apply) {
```
yeah good type change... I agree that hiding those UnitPFBuilder is good

jypma replied:
Are we really OK with `type` and `apply` not actually needing to be the same type? It allows for subtle compile errors. If we do I'd at least prefer a warning in the docs. To wrap up my earlier research:
- `P` is not bound to anything, so java is free to pick `Object`
- The `? extends P` was introduced to allow matching on the raw class of a generic message, e.g. `.match(List.class, (List<String> msg) -> ...)`
- However, this actually disconnects the two types completely, so one could have 
```
void handle(Msg1 m) { ... }
void handle(Msg2 m) { ... }
receiveBuilder.match(Msg1.class, this::handle)
```
and it wouldn't compile, since both `Msg1` and `Msg2` match `? extends Object`.

Similarly, you can do `.match(String.class, (Integer i) -> ...)` which would compile fine, but throw `ClassCastException` at runtime.

I've suggested the following before, I think after discussion with @ktoso:
1. Rename the existing method to `.matchUnchecked`, and simply declare both types `<?>` instead of `<? extends P>`
2. Have a new, actually type-safe method `<P> ReceiveBuilder match (final Class<P> type, final FI.UnitApply<P> apply)`

This comment also applies to `match` on `? extends P` with predicate.

patriknw replied:
Thanks for your insights. 
`match(List.class, (List<String> msg) -> ...)` doesn't sound like a good thing to do.
I think we should fix it as you suggest.

There are also other builders (not `ReceiveBuilder`) that are used in stable APIs such as Akka Streams `collect`. I'm more worried about changing those and break binary compatibility of Akka Streams. Have to look into that part.

patriknw commented:
@jypma Thanks for feedback. I consider you as power user, but we should support `PartialFunction` also. You can override `receive` and return the `PartialFunction` instead of using `initialReceive` (well you must implement it and throw `UnsupportedOperationException` or something).

You are almost better off using `Actor` than `AbstractActor` :)
Some type conversions in `AbstractActorContext` though.

ktoso commented on akka-actor-tests/src/test/java/akka/event/ActorWithMDC.java:
```diff
@@ -1,47 +1,50 @@
-                break;
-        }
+  @Override
+  public ReceiveBuilder initialReceive() {
```
We could make `Receive` type extending `PartialFunction` which would be a bit nicer.

ktoso commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -46,29 +50,97 @@ object AbstractActor {
+  /**
+   * Returns the ActorRef for this actor.
+   */
+  def getSelf(): ActorRef = self
```
So this is mostly for source compatibility with UntypedActor.
Could we deprecate them right away then? Since it would be nice to have 1 API for those, it should be the `sender()` and `self()` ones.

patriknw commented:
Pushed two changes.

`AbstractActor.Receive` is now the return type of `initialReceive`. It's simply a wrapper of the `PartialFunction` to make the signatures look nice.

I tried inheritance (Receive extends PartialFunction) but it goes into a complexity rabbit hole when trying to implement `orElse`. It would also have bad performance since `applyOrElse` is optimized.

I also moved AbstractActorContext to AbstractActor.ActorContext, since `AbstractActorContext` is a rather strange name. Should not break much source since it's not supposed to be passed around.

ktoso commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -23,6 +23,46 @@ object AbstractActor {
+   * The actor context - the view of the actor cell from the actor.
+   * Exposes contextual information for the actor and the current message.
+   */
+  trait ActorContext extends akka.actor.ActorContext {
```
good idea.

ktoso commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -23,6 +23,46 @@ object AbstractActor {
+     * leaks in case client code is written without consulting this documentation first (i.e.
+     * always pushing new behaviors and never issuing an `unbecome()`)
+     */
+    def become(behavior: Receive, discardOld: Boolean): Unit =
```
👍

ktoso commented on akka-actor-tests/src/test/java/akka/actor/ActorCreationTest.java:
```diff
@@ -96,8 +96,8 @@ public static Props propsUsingLamdaWithoutClass(Integer magicNumber) {
-      return receiveBuilder();
+    public Receive initialReceive() {
+      return receiveBuilder().build();
     }
```
Excellent 👍

ktoso commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -14,10 +14,15 @@ import scala.runtime.BoxedUnit
  */
 object AbstractActor {
+
+  class Receive(val onMessage: PartialFunction[Any, BoxedUnit]) {
```
final? 
was somewhat tempted to allow aroundReceive here but let's not...

ktoso replied:
Needs some docs what it is :)

ktoso commented:
Looking good. I guess we can then collapse docs and use this style in all samples?

jypma commented:
What do you guys think about the [matcher DSL in javaslang](http://www.javaslang.io/javaslang-docs/#_pattern_matching)? They seem to have struck a nice balance. They model partial functions as `T => Option[U]` though.

ktoso commented:
The purpose of this design is to allow someone to provide an API that would take / use the Slang matcher if anyone wants to.

jypma commented:
@ktoso `Receive` itself is final, is `ReceiveBuilder` supposed to be extended so one can hack around on `PartialFunction` directly? That would be needed, since in order to deal with `Option[T]` one would need to catch the `.isDefinedAt` invocation, and then hope that `.apply` is actually called immediately after :)

patriknw commented:
@jypma I'm not sure I understand what you need to do, but the intention with `Receive` is that it's a simple wrapper around a `PartialFunction`. If you can create a `PartialFunction in some way you can return it, wrapped in `Receive`. In other words, you don't have to use the `ReceiveBuilder` to create the `PartialFunction`.

jypma commented:
ah right. I missed that. Thanks!

patriknw commented:
@jypma I think it would be very interesting to support javaslang as a way of building the `Receive` (as alternative to `ReceiveBuilder`). We can't add dependency to javaslang in akka-actor but depending on how complicated the integration is it could end up as documentation or a separate module/library. Would you be interested in creating a prototype of that?

jypma commented:
@patriknw just did: https://github.com/Tradeshift/ts-reaktive/pull/50 . Seems to work OK, if a bit hacky with the thread local. Will keep working just fine after this PR, we'd just pre-wrap the custom `PartialFunction` in the new (akka) `Receive` class.

patriknw commented:
Refs https://github.com/akka/akka/issues/21717

patriknw commented:
Good progress with many hours of work. The reward is:

```
removing UntypedActor docs:
  63 files changed, 304 insertions(+), 7136 deletions(-)
```

Shouldn't be too much left now. Some more `UntypedActor` to convert in the documentation, and writing migration guide. I will do that next week. You can wait with reviewing until everything is completed.

patriknw commented on akka-actor/src/main/java/akka/japi/pf/ReceiveBuilder.java:
```diff
@@ -30,80 +35,164 @@
  *
- * This is an EXPERIMENTAL feature and is subject to change until it has received more real world testing.
  */
 public class ReceiveBuilder {
```
PLS REVIEW

patriknw commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -5,17 +5,74 @@
- *
- * This is an EXPERIMENTAL feature and is subject to change until it has received more real world testing.
  */
 object AbstractActor {
```
PLS REVIEW all of this file

patriknw commented on akka-actor/src/main/scala/akka/actor/ActorCell.scala:
```diff
@@ -377,7 +361,7 @@ private[akka] class ActorCell(
   val dispatcher:  MessageDispatcher,
   val parent:      InternalActorRef)
-  extends UntypedActorContext with AbstractActorContext with Cell
+  extends UntypedActorContext with AbstractActor.ActorContext with Cell
```
PLS REVIEW

patriknw commented on akka-docs/rst/java/code/docs/actorlambda/ActorDocTest.java:
```diff
@@ -71,49 +88,111 @@ public static void beforeClass() {
 
   static
   //#context-actorOf
   public class FirstActor extends AbstractActor {
```
PLS REVIEW this file

patriknw commented on akka-docs/rst/java/lambda-actors.rst:
```diff
@@ -615,33 +606,59 @@ routers, load-balancers, replicators etc.
 
+.. _actors-receive-java:
+
 Receive messages
```
PLS REVIEW this section

patriknw commented on akka-docs/rst/project/migration-guide-2.4.x-2.5.x.rst:
```diff
@@ -4,8 +4,195 @@
 
-Actor
-=====
+Actor (Java)
```
PLS REVIEW

patriknw commented on akka-persistence/src/main/scala/akka/persistence/PersistentActor.scala:
```diff
@@ -309,6 +310,37 @@ abstract class UntypedPersistentActor extends UntypedActor with Eventsourced wit
 abstract class AbstractPersistentActor extends AbstractActor with PersistentActor with Eventsourced {
 
   /**
```
PLS REVIEW

Note the naming of these methods, `defineReceiveRecover` and `defineReceiveCommand`.
I don't think it's consistent with `initalReceive`, but that might not matter. wdyt?
An alternative is to use `defineReceive` instead of `initialReceive`?

ktoso replied:
define is a bit weird... `initialRecoveryReceive` would be another option? Sounds more close to what we have elsewhere

patriknw reviewed (commented):
Please focus the review on the parts I have commented with `PLS REVIEW`.

patriknw commented:
This is ready for final review. I have added some `PLS REVIEW` comments in the diff so that you can focus on the most important.

There are 45 `UntypedActor` remaining to be converted, but that is just mechanical work. I have already converted 150.

I disabled samples from the build. Better to convert those in the new repository.

patriknw commented on akka-docs/rst/java/lambda-actors.rst:
```diff
@@ -1,8 +1,8 @@
-.. _lambda-actors-java:
+.. _actors-java:
 
```
Before merging I will also rename those doc files (remove the lambda prefix). Kept it now to not mess up the diff.

johanandren commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -29,57 +86,138 @@ object AbstractActor {
+   * The ActorContext is not thread safe so do not expose it outside of the
+   * AbstractActor.
+   */
+  def getContext(): AbstractActor.ActorContext = context.asInstanceOf[AbstractActor.ActorContext]
```
Should we add a "Scala API" on context in Actor as well to help java devs not choosing that one? (might be applicable for more accessor-like methods on Actor as well)

johanandren commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -29,57 +86,138 @@ object AbstractActor {
+  final def receiveBuilder(): ReceiveBuilder = ReceiveBuilder.create()
+}
+
+abstract class UntypedAbstractActor extends AbstractActor {
```
Doc on the class to highlight why it exists and pointing out that directly implementing `AbstractActor` is better?

johanandren reviewed (commented):
LGTM with some possible doco improvements noted.

jypma commented:
@patriknw Would it be OK to mention how to do javaslang receive's in the docs? Or is it a bit too "far out"? I only stumbled onto javaslang recently, I don't know how widespread its acceptance is in the "plain Java" crowd.

patriknw commented:
@jypma I think it would be interesting to have a section in the docs for javaslang. Can you contribute that when this has been merged?

jypma commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -5,17 +5,74 @@
+   * how the messages should be processed. You can build such behavior with the
+   * [[akka.japi.pf.receivebuilder]] but it can be implemented in other ways than
+   * using the `ReceiveBuilder` since it in the end is just a wrapper around a
+   * Scala `PartialFunction`.
```
Consider adding: `In Java, you can implement PartialFunction by extending AbstractPartialFunction.`

jypma commented:
Sure thing.

patriknw commented:
@jypma Please review that I got it right: 5a8c66a

jypma commented:
@patriknw Really thrilled to see the type system improvements!

patriknw commented on akka-actor/src/main/scala/akka/actor/AbstractActor.scala:
```diff
@@ -5,17 +5,74 @@
+     * Returns a reference to the named child or null if no child with
+     * that name exists.
+     */
+    def getChild(name: String): ActorRef
```
@johanandren Here is the one that would benefit from returning `Optional`. Don't know if that would be a scary change since the compiler might not catch it in all cases, e.g.

```java
  if (getContext().getChild("foo") != null)
    getContext().actorOf(Props..., "foo");
```

I would prefer deprecation and adding a new method, but all good names are taken.

jypma replied:
It's a scary one indeed. OTOH, it would be a really nice cleanup. Personally, I think I would bite the bullet and change it (if we can w.r.t. non-experimental API promises).

Maybe do a little poll on the chat?

johanandren replied:
I don't think it is worth the risk here.

patriknw replied:
`ActorIdentity.getRef` is similar, and we can't change that without deprecation. I think we have to come up with an alternative naming. `childOptional`, `getChildOptional`?

guidomedina replied:
I believe chances are of someone getting a child are because most likely is going to be present hence I wouldn't use optional if the deprecated alternative doesn't, what about:
`getChildIfPresent`

jlprat replied:
What about `locateChild` or `retrieveChild`

jypma replied:
or `existingChild`

ktoso replied:
`findChild`? Since new JDK APIs use `find` in methods that return optionals (I wonder which language they got that from ;o))

Example https://docs.oracle.com/javase/8/docs/api/java/util/stream/IntStream.html

patriknw replied:
such prefixes are hard to guess (IDE completion)

patriknw replied:
I like `findChild`, but it would not work for `ActorIdentity.ref`.
We can use it here though.

patriknw commented on project/MiMa.scala:
```diff
@@ -87,6 +87,19 @@ object MiMa extends AutoPlugin {
+      ProblemFilters.exclude[MissingClassProblem]("akka.actor.AbstractActorContext"),
+      ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.actor.AbstractActor.getContext"),
+      ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.actor.AbstractActor.emptyBehavior"),
+      ProblemFilters.exclude[MissingTypesProblem]("akka.actor.ActorCell"),
```
I'm somewhat scared of this change, that is due to the move of AbstractActorContext. Could it break bin compat for scala actors. I think I'll have to do a real test with an example application.

patriknw commented:
@johanandren @ktoso @jypma I have incorporated your feedback except from the `? super P`, see last 4 commits.

@jypma regarding `? super P` I could not reproduce the "problem", see new test in 471b3b0

johanandren commented on akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala:
```diff
@@ -128,7 +128,7 @@ object GraphStageLogic {
     materializer:     ActorMaterializer,
     getAsyncCallback: StageActorRef.Receive ⇒ AsyncCallback[(ActorRef, Any)],
-    initialReceive:   StageActorRef.Receive) {
+    createReceive:    StageActorRef.Receive) {
```
Should this really have been renamed, its not a lambda but a receive that was already created, keep old name?

patriknw replied:
good catch, it was replaced by the global find-replace

patriknw replied:
fixed

johanandren reviewed (approved):
LGTM except the one (accidental?) rename in `GraphStage`

patriknw committed:
rename lambda doc files, #21717

patriknw commented:
Renamed the `lamda-*.rst` files in docs. Might be good to keep that as a separate commit so that git understands that it's a rename.

patriknw commented:
PLS BUILD
